File: let/input/prescribe.py
"""
prescribe.py
============

Model
-----
Independent diseases
    P(flu)   = 0.10
    P(covid) = 0.25

Symptom generation (noisy-OR)
    P(cough | flu)   = 0.40     P(fever | flu)   = 0.20
    P(cough | covid) = 0.90     P(fever | covid) = 0.80

Drug effectiveness (noisy-OR of disease-conditioned causes)
    oseltamivir : 0.85 if flu   , 0.10 if covid
    paxlovid    : 0.05 if flu   , 0.90 if covid

Side-effect rates (independent)
    P(side_effect | oseltamivir) = 0.15
    P(side_effect | paxlovid)    = 0.25

good(D)  ≜  effective(D) ∧ ¬side_effect(D)

Function
--------
    prescribe(e_cough: bool, e_fever: bool)
        → {"oseltamivir": p, "paxlovid": p}
returns P(good(D) | evidence) for each drug.
"""

from itertools import product
from typing import Dict, Tuple


# ───────────────────────────────────────────────────────────────
# 1 ▸  Model parameters
# ───────────────────────────────────────────────────────────────
P_FLU     = 0.10
P_COVID   = 0.25

SYMPTOMS = {
    "cough": {"flu": 0.40, "covid": 0.90},
    "fever": {"flu": 0.20, "covid": 0.80},
}

EFFECTIVE = {
    "oseltamivir": {"flu": 0.85, "covid": 0.10},
    "paxlovid":    {"flu": 0.05, "covid": 0.90},
}

SIDE_EFFECT = {
    "oseltamivir": 0.15,
    "paxlovid":    0.25,
}


# ───────────────────────────────────────────────────────────────
# 2 ▸  Helper utilities
# ───────────────────────────────────────────────────────────────
def noisy_or(triggers) -> float:
    """Return 1 − ∏(1−p_i) for a list of probabilities p_i."""
    prod = 1.0
    for p in triggers:
        prod *= 1.0 - p
    return 1.0 - prod


def symptom_prob(symptom: str, flu: bool, covid: bool) -> float:
    """P(symptom | flu?, covid?) via noisy-OR of active causes."""
    triggers = []
    if flu:
        triggers.append(SYMPTOMS[symptom]["flu"])
    if covid:
        triggers.append(SYMPTOMS[symptom]["covid"])
    return noisy_or(triggers) if triggers else 0.0


def effective_prob(drug: str, flu: bool, covid: bool) -> float:
    """P(effective(drug) | flu?, covid?)"""
    triggers = []
    if flu:
        triggers.append(EFFECTIVE[drug]["flu"])
    if covid:
        triggers.append(EFFECTIVE[drug]["covid"])
    return noisy_or(triggers) if triggers else 0.0


# ───────────────────────────────────────────────────────────────
# 3 ▸  Exact Bayesian inference by enumeration
# ───────────────────────────────────────────────────────────────
def prescribe(evidence_cough: bool, evidence_fever: bool) -> Dict[str, float]:
    """
    Return posterior probability of good(drug) for each drug,
    conditioning on observed cough / fever.
    """
    worlds = []        # list of (flu?, covid?, joint weight P(w|e))
    for flu, covid in product([False, True], repeat=2):
        # prior over diseases (independent)
        p_prior = (P_FLU if flu else 1 - P_FLU) * \
                  (P_COVID if covid else 1 - P_COVID)

        # likelihood of observed symptoms
        p_cough = symptom_prob("cough", flu, covid)
        p_fever = symptom_prob("fever", flu, covid)

        p_lik  = (p_cough if evidence_cough else 1 - p_cough) * \
                 (p_fever if evidence_fever else 1 - p_fever)

        worlds.append((flu, covid, p_prior * p_lik))

    norm = sum(w[2] for w in worlds)             # P(evidence)

    post_good: Dict[str, float] = {d: 0.0 for d in EFFECTIVE}
    for drug in EFFECTIVE:
        p_se = SIDE_EFFECT[drug]

        for flu, covid, weight in worlds:
            p_eff  = effective_prob(drug, flu, covid)
            p_good = p_eff * (1 - p_se)          # independent side-effect
            post_good[drug] += weight * p_good

        post_good[drug] /= norm                  # normalise

    return post_good


# ───────────────────────────────────────────────────────────────
# 4 ▸  Demo
# ───────────────────────────────────────────────────────────────
if __name__ == "__main__":
    # Example: patient presents with both cough **and** fever
    posterior = prescribe(True, True)

    # pretty-print sorted by probability
    for drug, prob in sorted(posterior.items(), key=lambda kv: kv[1], reverse=True):
        print(f"Posterior P(good({drug})) = {prob:.3f}")

    best_drug, best_prob = max(posterior.items(), key=lambda kv: kv[1])
    print(f"\n--> Recommend {best_drug} "
          f"(probability of net benefit ≈ {best_prob:.1%})")

File: eyelet/input/prescribe.c
/*
 * prescribe.c  —  probabilistic drug-selection demo (plain C version)
 *
 * Compile:   gcc -O2 -std=c11 -o prescribe prescribe.c -lm
 * Run:       ./prescribe            # prints recommendation for cough+fever case
 *
 * The program reproduces the ProbLog model by exhaustive enumeration of the
 * four possible disease states (flu ∈{0,1}, covid ∈{0,1}) and then:
 *   1. computes the posterior P(flu, covid | evidence),
 *   2. marginalises to obtain P(good(oseltamivir)) and P(good(paxlovid)),
 *   3. prints the two posterior probabilities, sorted, and recommends the
 *      drug with the higher probability of *net benefit*.
 */

#include <stdio.h>
#include <stdbool.h>

/* ---------- fixed model parameters ---------- */
enum {   /* indices for disease combinations (flu<<1 | covid) */
    NONE = 0, FLU_ONLY = 2, COVID_ONLY = 1, BOTH = 3
};

/* Priors */
static const double P_FLU   = 0.10;
static const double P_COVID = 0.25;

/* Symptom generation */
static const double P_COUGH_FLU   = 0.40;
static const double P_COUGH_COVID = 0.90;
static const double P_FEVER_FLU   = 0.20;
static const double P_FEVER_COVID = 0.80;

/* Drug effectiveness */
static const double P_EFF_OSE_FLU   = 0.85;
static const double P_EFF_OSE_COVID = 0.10;
static const double P_EFF_PAX_COVID = 0.90;
static const double P_EFF_PAX_FLU   = 0.05;

/* Side-effect risks (independent of disease) */
static const double P_SE_OSE = 0.15;
static const double P_SE_PAX = 0.25;

/* ---------- helpers ---------- */

/* P(symptom | flu, covid)  —  independent noisy-OR */
static double
p_symptom(bool flu, bool covid, double p_given_flu, double p_given_covid)
{
    double p_no = 1.0;
    if (flu)   p_no *= (1.0 - p_given_flu);
    if (covid) p_no *= (1.0 - p_given_covid);
    return 1.0 - p_no;
}

/* P(effective(drug) | flu, covid)  —  independent noisy-OR */
static double
p_effective_ose(bool flu, bool covid)
{
    double p_no = 1.0;
    if (flu)   p_no *= (1.0 - P_EFF_OSE_FLU);
    if (covid) p_no *= (1.0 - P_EFF_OSE_COVID);
    return 1.0 - p_no;
}

static double
p_effective_pax(bool flu, bool covid)
{
    double p_no = 1.0;
    if (covid) p_no *= (1.0 - P_EFF_PAX_COVID);
    if (flu)   p_no *= (1.0 - P_EFF_PAX_FLU);
    return 1.0 - p_no;
}

/* ---------- core inference ---------- */
typedef struct {
    double p_good_ose;
    double p_good_pax;
} Posterior;

/* Returns P(good(oseltamivir)) and P(good(paxlovid)) given evidence */
static Posterior
prescribe(bool evidence_cough, bool evidence_fever)
{
    /* 1.  posterior over four disease states --------------------------- */
    double w[4] = {0};             /* unnormalised weights               */
    double norm = 0.0;

    for (int flu = 0; flu <= 1; ++flu) {
        for (int covid = 0; covid <= 1; ++covid) {
            double prior = (flu   ? P_FLU   : 1.0 - P_FLU) *
                           (covid ? P_COVID : 1.0 - P_COVID);

            double p_cough = p_symptom(flu, covid, P_COUGH_FLU, P_COUGH_COVID);
            double p_fever = p_symptom(flu, covid, P_FEVER_FLU, P_FEVER_COVID);

            double p_evidence =
                (evidence_cough ? p_cough : (1.0 - p_cough)) *
                (evidence_fever ? p_fever : (1.0 - p_fever));

            int idx = (flu << 1) | covid;  /* NONE, COVID_ONLY, FLU_ONLY, BOTH */
            w[idx]  = prior * p_evidence;
            norm   += w[idx];
        }
    }
    for (int i = 0; i < 4; ++i) w[i] /= norm;   /* normalise */

    /* 2.  marginalise to drug-level probabilities ---------------------- */
    double p_good_ose = 0.0, p_good_pax = 0.0;
    for (int flu = 0; flu <= 1; ++flu) {
        for (int covid = 0; covid <= 1; ++covid) {
            int idx = (flu << 1) | covid;

            double p_eff_ose = p_effective_ose(flu, covid);
            double p_eff_pax = p_effective_pax(flu, covid);

            double p_good_ose_state = p_eff_ose * (1.0 - P_SE_OSE);
            double p_good_pax_state = p_eff_pax * (1.0 - P_SE_PAX);

            p_good_ose += w[idx] * p_good_ose_state;
            p_good_pax += w[idx] * p_good_pax_state;
        }
    }
    return (Posterior){p_good_ose, p_good_pax};
}

/* ---------- demonstration (same as Python’s __main__) ----------------- */
int main(void)
{
    /* Example: patient presents with BOTH cough **and** fever */
    bool cough = true;
    bool fever = true;

    Posterior post = prescribe(cough, fever);

    /* pretty-print in descending order */
    if (post.p_good_ose > post.p_good_pax) {
        printf("Posterior P(good(oseltamivir)) = %.3f\n", post.p_good_ose);
        printf("Posterior P(good(paxlovid))    = %.3f\n", post.p_good_pax);
        printf("\n--> Recommend oseltamivir "
               "(probability of net benefit ≈ %.1f%%)\n",
               100.0 * post.p_good_ose);
    } else {
        printf("Posterior P(good(paxlovid))    = %.3f\n", post.p_good_pax);
        printf("Posterior P(good(oseltamivir)) = %.3f\n", post.p_good_ose);
        printf("\n--> Recommend paxlovid "
               "(probability of net benefit ≈ %.1f%%)\n",
               100.0 * post.p_good_pax);
    }
    return 0;
}

